code/modules/mob/living/silicon/pai/software_modules.dm

```diff
@@ -0,0 +1,539 @@
+		if(!ui)
+			// Don't copy-paste this unless you're making a pAI software module!
+			ui = new(user, user, id, "pai_directives.tmpl", "pAI Directives", 450, 600)
+			ui.set_initial_data(data)
+			ui.open()
+			ui.set_auto_update(1)
+
+	Topic(href, href_list)
+		var/mob/living/silicon/pai/P = usr
+		if(!istype(P)) return
+
+		if(href_list["getdna"])
+			var/mob/living/M = P.loc
+			var/count = 0
+
+			// Find the carrier
+			while(!istype(M, /mob/living))
+				if(!M || !M.loc || count > 6)
+					//For a runtime where M ends up in nullspace (similar to bluespace but less colourful)
+					src << "You are not being carried by anyone!"
+					return 0
+				M = M.loc
+				count++
+
+			// Check the carrier
+			var/answer = input(M, "[P] is requesting a DNA sample from you. Will you allow it to confirm your identity?", "[P] Check DNA", "No") in list("Yes", "No")
+			if(answer == "Yes")
+				var/turf/T = get_turf_or_move(P.loc)
+				for (var/mob/v in viewers(T))
+					v.show_message("<span class='notice'>[M] presses \his thumb against [P].</span>", 3, "<span class='notice'>[P] makes a sharp clicking sound as it extracts DNA material from [M].</span>", 2)
+				var/datum/dna/dna = M.dna
+				P << "<font color = red><h3>[M]'s UE string : [dna.unique_enzymes]</h3></font>"
+				if(dna.unique_enzymes == P.master_dna)
+					P << "<b>DNA is a match to stored Master DNA.</b>"
+				else
+					P << "<b>DNA does not match stored Master DNA.</b>"
+			else
+				P << "[M] does not seem like \he is going to provide a DNA sample willingly."
+			return 1
+
+/datum/pai_software/radio_config
+	name = "Radio Configuration"
+	ram_cost = 0
+	id = "radio"
+	toggle = 0
+	default = 1
+
+	on_ui_interact(mob/living/silicon/pai/user, datum/nanoui/ui = null, force_open = 1)
+		var/data[0]
+
+		data["listening"] = user.radio.broadcasting
+		data["frequency"] = format_frequency(user.radio.frequency)
+
+		var/channels[0]
+		for(var/ch_name in user.radio.channels)
+			var/ch_stat = user.radio.channels[ch_name]
+			var/ch_dat[0]
+			ch_dat["name"] = ch_name
+			// FREQ_LISTENING is const in /obj/item/device/radio
+			ch_dat["listening"] = !!(ch_stat & user.radio.FREQ_LISTENING)
+			channels[++channels.len] = ch_dat
+
+		data["channels"] = channels
+
+		ui = nanomanager.try_update_ui(user, user, id, ui, data, force_open)
+		if(!ui)
+			ui = new(user, user, id, "pai_radio.tmpl", "Radio Configuration", 300, 150)
+			ui.set_initial_data(data)
+			ui.open()
+
+	Topic(href, href_list)
+		var/mob/living/silicon/pai/P = usr
+		if(!istype(P)) return
+
+		P.radio.Topic(href, href_list)
+		return 1
+
+/datum/pai_software/crew_manifest
+	name = "Crew Manifest"
+	ram_cost = 5
+	id = "manifest"
+	toggle = 0
+
+	on_ui_interact(mob/living/silicon/pai/user, datum/nanoui/ui=null, force_open=1)
+
+		var/data[0]
+		data["manifest"] = data_core.get_manifest_json()
+		// This is dumb, but NanoUI breaks if it has no data to send
+		data["a"] = "a"
```

Comment:
This isn't needed here, I only added this because Baystation12's NanoUI cache meant all the data for the UI was cached and none needed sending as proper data, and NanoUI didn't like that. since the manifest is being sent as normal data here, the dummy data doesn't need adding.